modules/rest-compatibility/src/main/java/org/elasticsearch/rest/compat/version7/RestCreateIndexActionV7.java

```diff
@@ -0,0 +1,110 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.rest.compat.version7;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.admin.indices.create.CreateIndexRequest;
+import org.elasticsearch.action.support.ActiveShardCount;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.RestToXContentListener;
+import org.elasticsearch.rest.action.admin.indices.RestCreateIndexAction;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.util.Collections.singletonMap;
+
+public class RestCreateIndexActionV7 extends RestCreateIndexAction {
+
+    /**
+     * Parameter that controls whether certain REST apis should include type names in their requests or responses.
+     * Note: Support for this parameter will be removed after the transition period to typeless APIs.
+     */
+    public static final String INCLUDE_TYPE_NAME_PARAMETER = "include_type_name";
+    public static final boolean DEFAULT_INCLUDE_TYPE_NAME_POLICY = false;
+
+    @Override
+    public String compatibleWithVersion() {
+        return String.valueOf(Version.V_7_0_0.major);
+    }
+
+    @Override
+    public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
+        CreateIndexRequest createIndexRequest = new CreateIndexRequest(request.param("index"));
+
+        if (request.hasContent()) {
+            Map<String, Object> sourceAsMap = XContentHelper.convertToMap(request.requiredContent(), false, request.getXContentType()).v2();
+
+            request.param(INCLUDE_TYPE_NAME_PARAMETER);// just consume, it is not replaced with _doc
+            sourceAsMap = prepareMappingsV7(sourceAsMap, request);
+
+            createIndexRequest.source(sourceAsMap, LoggingDeprecationHandler.INSTANCE);
+        }
+
+        createIndexRequest.timeout(request.paramAsTime("timeout", createIndexRequest.timeout()));
+        createIndexRequest.masterNodeTimeout(request.paramAsTime("master_timeout", createIndexRequest.masterNodeTimeout()));
+        createIndexRequest.waitForActiveShards(ActiveShardCount.parseString(request.param("wait_for_active_shards")));
+        return channel -> client.admin().indices().create(createIndexRequest, new RestToXContentListener<>(channel));
+    }
+
+    static Map<String, Object> prepareMappingsV7(Map<String, Object> source, RestRequest request) {
+        final boolean includeTypeName = request.paramAsBoolean(INCLUDE_TYPE_NAME_PARAMETER, DEFAULT_INCLUDE_TYPE_NAME_POLICY);
+
+        @SuppressWarnings("unchecked")
+        Map<String, Object> mappings = (Map<String, Object>) source.get("mappings");
```

Comment:
good question.. I forgot that we could have multiple types. When was this deprecated? Do we expect multiple types in mappings in 7.x?
from what I see in `Internal CreateIndexRequest holds a single mapping (#48996) Alan Woodward* 18/11/2019, 11:52` it is aimed at 8. so we should expect multiple types in 7.0-7.x?

Reply:
good question.. I forgot that we could have multiple types. When was this deprecated? Do we expect multiple types in mappings in 7.x?
from what I see in Internal CreateIndexRequest holds a single mapping (#48996) Alan Woodward* 18/11/2019, 11:52 it is aimed at 8. so we should expect multiple types in 7.0-7.x?

Reply:
I confirmed this with @colings86 that multiple types were deprecated in 6.x and were not expected anymore. 
if it contains two types it will get into else branch here https://github.com/elastic/elasticsearch/pull/54197/files#diff-9327cfc566910ab3fec1c4b23a87a71bR90
for `v8 logic`. It does not expect two types either, and in fact will consider this mapping `not typed`
``` 
MapperService.java
public static boolean isMappingSourceTyped(String type, Map<String, Object> mapping) {
        return mapping.size() == 1 && mapping.keySet().iterator().next().equals(type);
    }
```
the request with two types in a mapping will  eventually end up with
```
     "type": "mapper_parsing_exception",
        "reason": "Root mapping definition has unsupported parameters:  [some_type_2 : {properties={field1={type=text}}}] [some_type_1 : {properties={field1={type=text}}}]"
      }
```
Should we improve this or is this good enough?

Reply:
thanks for following up. I think that error is good enough.